repo-docs/release-process.md

```diff
@@ -0,0 +1,113 @@
+# Release processes
+
+This repository contains two release processes, which are triggered on Git commits:
+
+## 1. The pre-release process
+
+The purpose of the pre-release process is to publish new changes from the main branch as frequently as possible. This has two main benefits:
+
+1. This tests whether we broke some parts of the release process so we don't get surprises when we need a manual release.
+2. This enables implementers to test a new feature or fix before we trigger a manual release.
+
+To achieve its purpose, the pre-release process is executed on every new commit on the main branch. Additionally, pre-releases do not commit anything to the main branch, which allows them to be deployed even if multiple features are merged on the main branch faster than the CI can release them.
+
+## 2. The manual release process
+
+The purpose of the manual release process is to deploy versions of our packages that we feel confident are safe for implementers to use.
+
+To achieve that purpose, the manual release process is triggered manually whenever we feel that the version currently on the main branch is safe. Additionally, every Wednesday, to ensure we don't forget to regularly trigger manual releases, a request for approval to deploy is triggered.
+
+Whenever a manual release is successful, a commit is pushed to the main branch which bumps the version of every publishable package in the repo.
+
+# Versioning & publishing to NPM
+
+Versions for any given commit are determined based on the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.
+
+Specifically, a commit will determine its version by looking for the last manual release version and bumping it based on how breaking the changes are between then and the current commit. This ensures that implementers can safely update their dependencies on our packages without unexpectedly causing errors.
+
+When triggered, releases processes will execute a series of [Nx tasks](https://nx.dev/core-features/run-tasks). Some tasks are run at the root of the repository, and some will be run on each individual package.
+
+## `release:phase0` (lock the main branch)
+
+This task is only run for the manual release.
+
+The purpose of this task is to lock the main branch, preventing users from merging pull requests of committing.
+
+Specifically, this prevents new fixes or features from getting merged between the current release and the version bump commit we're about to push. If fixes or features were merged before the upcoming version bump, they would not be taken into account when calculating versions for the next releases.
+
+This task accomplishes its purpose by:
+
+1. Committing a ".git-lock` file to the main branch.
+2. Updating the repository's settings to enable "Require branches to be up to date before merging" for the main branch.
```

Comment:
Broader question - why not have this enabled by default ?

Reply:
For most repos, I think it'd be fine, but our repo is a monorepo with multiple teams working on it and a lot of PRs open at once, and our CI still takes quite a few minutes to run. This would block people very often and be very annoying. We could investigate using merge queues, but I haven't looked much into it.